nbconvert/preprocessors/tests/test_execute.py

```diff
@@ -40,7 +42,64 @@ def _normalize_base64(b64_text):
     except (ValueError, TypeError):
         return b64_text
 
-class TestExecute(PreprocessorTestsBase):
+
+def merge_dicts(first, second):
+    # Because this is annoying to do inline
+    outcome = {}
+    outcome.update(first)
+    outcome.update(second)
+    return outcome
+
+
+class ExecuteTestBase(PreprocessorTestsBase):
+    def build_preprocessor(self, opts):
+        """Make an instance of a preprocessor"""
+        preprocessor = ExecutePreprocessor()
+        preprocessor.enabled = True
+        for opt in opts:
+            setattr(preprocessor, opt, opts[opt])
+        # Perform some state setup that should probably be in the init
+        preprocessor._display_id_map = {}
+        preprocessor.widget_state = {}
+        preprocessor.widget_buffers = {}
+        return preprocessor
+
+    @staticmethod
+    def prepare_cell_mocks(*messages):
+        messages = list(messages)
+        def prepared_wrapper(func):
+            @functools.wraps(func)
+            def test_mock_wrapper(self):
+                parent_id = 'fake_id'
+                cell_mock = MagicMock(source='"foo" = "bar"', outputs=[])
+                # Hack to help catch `.` and `[]` style access to outputs against the same mock object
+                cell_mock.__getitem__.side_effect = lambda n: cell_mock.outputs if n == 'outputs' else None
```

Comment:
is this to get IPython [`IPython.utils.ipstruct.Struct`](https://ipython.readthedocs.io/en/stable/api/generated/IPython.utils.ipstruct.html#IPython.utils.ipstruct.Struct) like functionality? Is this needed because the MagicMock isn't able to inherit from that? Could we create a MagicMock subclass that just builds this functionality in for any of its attributes if we're going to be treating it like a struct?